### DISCUSSION TITLE
[Proposal] Allowing Extension Method Invocation on Method Groups in C#

### DISCUSSION THREAD

**Discussion by JKamsker** [dotnet/csharplang#7256] [Language Ideas]
This proposal is about enhancing the C# language to allow direct invocation of extension methods on method groups. The objective is to simplify code and improve readability in scenarios where method groups and extension methods are used together.

For example, consider the following code snippet:

```csharp
public static async IAsyncEnumerable<T> EnumerateAllAsync<T>(this Func<int, int, Task<ListResponse<T>>> fetchPageAsync)
{
    // Implementation omitted for brevity
}

async Task<ListResponse<SomeResponse>> FetchSomePageAsync(int offset, int limit)
{
    // Implementation omitted for brevity
}

```

The following code snippet shows how the extension method can be invoked currently on a method group:

```csharp
var func = FetchSomePageAsync;
await func.EnumerateAllAsync();
```

The proposed enhancement would allow the following code snippet to be used instead:

```csharp
await FetchSomePageAsync.EnumerateAllAsync();
```

Enabling direct extension method invocation on method groups would eliminate the need for an intermediate variable assignment, resulting in cleaner and more expressive code. This enhancement aligns with existing method group conversion and promotes code consistency.

This proposal aims to enhance developer experience, improve code expressiveness, and simplify the usage of method groups with extension methods. It would contribute to a more elegant and concise coding style for C# developers.

**Comment by TahirAhmadov** [comment]
Not sure if I like this approach entirely. An extension method is supposed to do something which is inherent to the type being extended. I'd prefer a static helper method which takes `Func<int, int, Task<ListResponse<T>>>` as a parameter, purely from elegance perspective.

However, the proposal as written may still have merits, I just personally don't find the stated use case compelling.

**Comment by CyrusNajmabadi** [comment]
I personally find it odd that extension methods aren't always available to be used as an extension in cases where they could be used as a normal static method.  So i'm for this.  But i would extend to all cases where we disallow this today.

**Reply by HaloFour** [reply]
I guess this only works now that lambdas and method groups have natural types (that prefer `Action<...>`/`Func<...>`?
